controller.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	"cloud.google.com/go/compute/apiv1/computepb"
+	"github.com/googleapis/gax-go/v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type runtimeClient interface {
+	Get(ctx context.Context, key types.NamespacedName, obj client.Object, opts ...client.GetOption) error
+}
+
+type instanceClient interface {
+	SetSecurityPolicy(ctx context.Context, req *computepb.SetSecurityPolicyInstanceRequest, opts ...gax.CallOption) (*compute.Operation, error)
+}
+
+type reconcileNode struct {
+	// client can be used to retrieve objects from the APIServer.
+	client            runtimeClient
+	instanceClient    instanceClient
+	selector          labels.Selector
+	securityPolicyURL string
+	projectID         string
+}
+
+// Assert reconcileNode implements reconcile.Reconciler.
+var _ reconcile.Reconciler = &reconcileNode{}
+
+func (r *reconcileNode) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Fetch the Node from the cache
+	node := &corev1.Node{}
+	if err := r.client.Get(ctx, request.NamespacedName, node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not fetch Node: %+v", err)
```

Comment:
log here, similar message to the error. This is a pretty exceptional condition (I think? Unless we see deletions?), so a log.Error seems ok.

Reply:
according to the Reconciler interface definition, reconcile is triggered in deletions as well

https://github.com/kubernetes-sigs/controller-runtime/blob/1ed345090869edc4bd94fe220386cb7fa5df745f/pkg/reconcile/reconcile.go#L62-L63

also what's the difference between log.Error vs fmt.Errorf? I saw online there're discussions about thread-safety, is that what you are thinking about?

Reply:
`fmt.Errorf` returns a string, `log.Error` logs an error. To do both you need two separate statements.